### ISSUE TITLE
Add, update, and remove repository rules

### ISSUE THREAD

**Issue by jayconrod** [bazelbuild/bazel-gazelle#56]
This is a catch-all issues for Gazelle repository rule management.

Gazelle will support an `update-repos` command. This will be able to:

* Add, remove, or update a repository rule in WORKSPACE for a specific dependency.
* Add repository rules for direct and transitive dependencies, based on import paths from Go libraries.
* Update a set of repository rules (possibly all).
* Remove repository rules for dependencies that aren't needed anymore.

**Comment by ashi009** [comment]
I guess this will not come before vgo hits GA?

**Comment by ashi009** [comment]
Scratch GA. After GA, I think it will take a while before the vgo
versioning system is “generally adopted” ;-)
Jay Conrod <notifications@github.com>于2018年6月28日 周四下午10:24写道：

**Comment by clintberry** [comment]
The link to the tool that adds repo rules to WORKSPACE doesn't appear to work anymore. The link is in this comment: https://github.com/bazelbuild/rules_go/issues/295#issuecomment-285388447

What is the latest and greatest way to update the WORKSPACE with all my external dependencies automagically? :-)

**Comment by linzhp** [comment]
To get the transitive dependencies of an external package, I am thinking about the following:

1. Create a Go module in TempDir
2. Convert the existing `go_repository` rules into the new go.mod file in TempDir.
3. Run "go get <import path>@vx.x.x" in the TempDir
4. Add/update existing `go_repository` rules based on the new go.mod/go.sum files

I don't want to assume `go.mod` file exists in the main repo, and instead use `go_repository` rules as the source of truth in Step 2.

@jayconrod What do you think?

**Comment by jayconrod** [comment]
@linzhp I'm not sure the `go` command should be involved at all in this process. There may be a number of workspaces declared with `http_archive` or other repository rules, and `go get` won't be able to work with those. Gazelle won't be able to update those, but it should still be able to load dependencies from them.

Some more thought also needs to go into making this language agnostic.

**Comment by linzhp** [comment]
I agree that we should make this language agnostic by having each language extension implement how they get their dependencies and resolve the versions. But for Go, I don't see a big issue with the "go get" approach:

- With close to 1000 external libraries we are using in our Go repo, we don't see any other Go library except Gazelle itself for which we cannot use `go_repository`. We even use `go_repository` for some repo that only contains proto files and use them to generate Go code at build time.
- Gazelle only generates `go_repository` rules for Go libraries. I think it's a fair assumption that if people decide to use `http_archive` or other workspace rules to download their Go libraries, they will have to manage those rules themselves.
- With no public API available, `go get` seems the only way we can resolve the transitive deps and their version in a consistent way to the Go modules system. Gazelle calls other go subcommands anyway.
